relayer/middleware/missedVaasV3/storage.ts

```diff
@@ -0,0 +1,330 @@
+  );
+
+  let failedToFetchSequences;
+
+  try {
+    failedToFetchSequences = await getDataFromSortedSet(
+      redis,
+      failedToFetchKey,
+    );
+  } catch (error) {
+    return error;
+  }
+
+  return failedToFetchSequences;
+}
+
+export async function calculateStartingIndex(
+  redis: Redis | Cluster,
+  prefix: string,
+  emitterChain: number,
+  emitterAddress: string,
+  lastSafeSequence?: bigint,
+  startingSequence?: bigint,
+  logger?: Logger,
+) {
+  const key = getSeenVaaKey(prefix, emitterChain, emitterAddress);
+
+  let indexToStartFrom: number;
+
+  if (!lastSafeSequence && startingSequence) {
+    indexToStartFrom = await redis.zrank(key, startingSequence.toString());
+    if (!indexToStartFrom) {
+      logger.warn(
+        "Starting Sequence Config not found in redis. No starting sequence will be used.",
+      );
+    }
+  } else if (lastSafeSequence) {
+    indexToStartFrom = await redis.zrank(key, lastSafeSequence.toString());
+  }
+
+  return indexToStartFrom;
+}
+
+export async function getAllProcessedSeqsInOrder(
+  redis: Redis | Cluster,
+  prefix: string,
+  emitterChain: number,
+  emitterAddress: string,
+  indexToStartFrom?: number,
+): Promise<bigint[]> {
+  const key = getSeenVaaKey(prefix, emitterChain, emitterAddress);
+
+  const results = await getDataFromSortedSet(
+    redis,
+    key,
+    indexToStartFrom?.toString(),
+  );
+  return results
+    .map(r => Number(r))
+    .sort((a, b) => a - b)
+    .map(BigInt);
+}
+
+/**
+ *
+ * Private Functions:
+ *
+ */
+
+function getSeenVaaKey(
+  prefix: string,
+  emitterChain: number,
+  emitterAddress: string,
+): string {
+  return `${prefix}:missedVaasV3:seenVaas:${emitterChain}:${emitterAddress}`;
+}
+
+function getFailedToFetchKey(
+  prefix: string,
+  emitterChain: number,
+  emitterAddress: string,
+): string {
+  return `${prefix}:missedVaasV3:failedToFetch:${emitterChain}:${emitterAddress}`;
+}
+
+function getSafeSequenceKey(
+  prefix: string,
+  emitterChain: number,
+  emitterAddress: string,
+): string {
+  return `${prefix}:missedVaasV3:safeSequence:${emitterChain}:${emitterAddress}`;
+}
+
+// example keys:
+// {GenericRelayer}:GenericRelayer-relays:14/000000000000000000000000306b68267deb7c5dfcda3619e22e9ca39c374f84/55/O8xvv:logs
+// {GenericRelayer}:GenericRelayer-relays:14/000000000000000000000000306b68267deb7c5dfcda3619e22e9ca39c374f84/49/d1Cjd
+function parseStorageVaaKey(key: string) {
+  const vaaIdString = key.split(":")[2];
+  const sequenceString = vaaIdString.split("/")[2];
+  return sequenceString;
+}
+
+const CURSOR_IDENTIFIER = "0";
+async function scanNextBatchAndUpdateSeenSequences(
+  redis: Redis | Cluster,
+  filter: FilterIdentifier,
+  storagePrefix: string,
+  seenVaaKey: string,
+  cursor: string = CURSOR_IDENTIFIER,
+): Promise<number> {
+  const { emitterChain, emitterAddress } = filter;
+  const prefix = `${storagePrefix}:${emitterChain}/${emitterAddress}`;
+  const [nextCursor, keysFound] = await redis.scan(
```

Comment:
Who owns the keys we are scanning here?

Reply:
they belong to redis-storage. This is creating an unwanted dependency with the RedisStorage, which is only one of the possible implementations of the storage interface.
This is definetly a tech debt and we'll create an issue accordingly.
I considered this tech debt to be acceptable for the moment because most many of the middleware/tools we are using have a dependency on redis-storage. I've added this comment about this on the config that enables this behaviour:

https://github.com/wormhole-foundation/relayer-engine/pull/185/files#diff-fe7adf976497ce5d7eb9e295f6478bdf8f063ba1791187bc45a80e041fdb03f3R50

I've also added a ticket to refactor this to our backlog, we can prioritize it in upcoming refinement/planning sessions

Reply:
I also tried to make the dependency explicit here:
https://github.com/wormhole-foundation/relayer-engine/pull/185/files#diff-01c3f5d8b2e9eebd1441f6da69a257153c8fc44933f793e6b2689378263346b1R112